zerver/middleware.py

```diff
@@ -306,14 +306,16 @@ def __call__(self, request: HttpRequest) -> HttpResponse:
             unset_request()
 
 
-def parse_client(request: HttpRequest) -> Tuple[str, Optional[str]]:
+@has_request_variables
```

Comment:
Yeah, I'd add a comment here explaining that we're taking advantage of has_request_variables supporting being called multiple times when processing a given request.

One thing we should do is test what happens if we throw a JsonableError in a validator in this code path -- which is basically a question of how the `process_exception` middleware interacts with an exception here.  There may be some small adjustments we want to make to make sure we get a "nice" error outcome in that case.

Reply:
Oh, I was imagining just adding a `raise JsonableError` here and then manually testing with loading Zulip in a browser to see what happens. We may or may not feel the need for an automated test depending on what we find.

Reply:
I did this manual test real quick, and wrote a commit you can squash that I think addresses it. I think this will fail test coverage, and that's probably worth fixing. I suspect one can write a fairly normal test that just uses `with mock.patch("zerver.lib.user_agent.parse_user_agent", side_effect=JsonableError("Oh no!"):` or something along those lines to test the new code path.

Reply:
Okay. Thanks for that clarification and doing that. I've added a test and squashed the commits. This was my first time using `mock.patch(...)` in a test (though I've spent a decent amount of time reading through tests that use it), so it would be worth a review to make sure I didn't miss anything. I couldn't get it to work using `zerver.lib.user_agent.parse_user_agent` but rather used `zerver.middleware.parse_client`.

Reply:
Cool, yeah, you did this correctly; `mock.patch` is a little finicky, but often you need to patch it in the namespace where it's imported to be called, which is indeed `zerver.middelware` :)